perf(core): make sanitization tree-shakable in Ivy mode

mhevery opened the pull request:
In VE the `Sanitizer` is always available in `BrowserModule` because the VE retrieves it using injection. 

In Ivy the injection is optional and we have instructions instead of component definition arrays. The implication of this is that in Ivy the instructions can pull in the sanitizer only when they are working with a property which is known to be unsafe. Because the Injection is optional this works even if no Sanitizer is present. So in Ivy we first use the sanitizer which is pulled in by the instruction, unless one is available through the `Injector` then we use that one instead.

This PR does few things:
1) It makes `Sanitizer` optional in Ivy.
2) It makes `DomSanitizer` tree shakable. 
3) It aligns the semantics of Ivy `Sanitizer` with that of the Ivy sanitization rules.
4) It refactors `DomSanitizer` to use same functions as Ivy sanitization for consistency.

IgorMinar commented on packages/core/src/sanitization/bypass.ts:
```diff
@@ -6,61 +8,113 @@
-  ResourceUrl = 'ResourceUrl',
+  Url = 'URL',
+  Html = 'HTML',
+  ResourceUrl = 'ResourceURL',
```
were these changes necessary or are they just stylistic?

mhevery replied:
Necessary to align the Ivy sanitization with `DomSanitizer`.

IgorMinar commented on packages/platform-browser/src/browser.ts:
```diff
@@ -37,9 +37,13 @@ export const INTERNAL_BROWSER_PLATFORM_PROVIDERS: StaticProvider[] = [
 ];
 
+export const BROWSER_SANITIZATION_PROVIDERS__POST_R3__ = [];
+export const BROWSER_SANITIZATION_PROVIDERS__PRE_R3__ = BROWSER_SANITIZATION_PROVIDERS;
```
is this right? I thought the right way to do this was to assign `{provide: Sanitizer, useExisting: DomSanitizer}` to `BROWSER_SANITIZATION_PROVIDERS__PRE_R3__` and then assign `BROWSER_SANITIZATION_PROVIDERS__PRE_R3__` to `BROWSER_SANITIZATION_PROVIDERS`

IgorMinar replied:
you seem to be doing it backwards

mhevery replied:
Yeah, sorry fixed.

IgorMinar commented on packages/platform-browser/src/browser.ts:
```diff
@@ -37,9 +37,13 @@ export const INTERNAL_BROWSER_PLATFORM_PROVIDERS: StaticProvider[] = [
+export const BROWSER_SANITIZATION_PROVIDERS__POST_R3__ = [];
+export const BROWSER_SANITIZATION_PROVIDERS__PRE_R3__ = BROWSER_SANITIZATION_PROVIDERS;
+
+const _BROWSER_SANITIZATION_PROVIDERS_INTERNAL = BROWSER_SANITIZATION_PROVIDERS__PRE_R3__;
```
if you do it the way I suggested there is no need for this const

mhevery replied:
removed

IgorMinar commented on packages/platform-browser/src/private_export.ts:
```diff
@@ -20,3 +20,9 @@ export {HammerGesturesPlugin as ɵHammerGesturesPlugin} from './dom/events/hamme
+// performed by rollup while it's creating fesm files.
+//
+// no code actually imports these symbols from the @angular/platform-browser entry point
+export {BROWSER_SANITIZATION_PROVIDERS__POST_R3__} from './browser';
```
won't this still make it part of the public api unless you prefix it with ɵ?

mhevery replied:
correct, fixed.

IgorMinar commented on packages/platform-browser/test/browser/bootstrap_spec.ts:
```diff
@@ -189,6 +189,19 @@ function bootstrap(
+           // instructions as needed.
+           expect(sanitizer).toBe(null);
+         } else {
+           // In VE we always need to have sanitize available.
```
```suggestion
           // In VE we always need to have Sanitizer available.
```

mhevery replied:
fixed

IgorMinar reviewed (commented):
I need to review this more carefully to have good feedback, so I'm just posting initial observations.

Three more things:
- Can you please write up a good commit message describing the change and providing enough context for all the changes you made?
- Can you take a look at this older PR that Martin has been trying to land for a while: https://github.com/angular/angular/pull/12310 and see if it makes sense to incorporate it into this change?
- We will likely need Rafael or someone else from the security team to take a look at this as well as the changes you made are nontrivial.

mhevery committed:
perf(core): make sanitization tree-shakable in Ivy mode

In VE the `Sanitizer` is always available in `BrowserModule` because the VE retrieves it using injection.

In Ivy the injection is optional and we have instructions instead of component definition arrays. The implication of this is that in Ivy the instructions can pull in the sanitizer only when they are working with a property which is known to be unsafe. Because the Injection is optional this works even if no Sanitizer is present. So in Ivy we first use the sanitizer which is pulled in by the instruction, unless one is available through the `Injector` then we use that one instead.

This PR does few things:
1) It makes `Sanitizer` optional in Ivy.
2) It makes `DomSanitizer` tree shakable.
3) It aligns the semantics of Ivy `Sanitizer` with that of the Ivy sanitization rules.
4) It refactors `DomSanitizer` to use same functions as Ivy sanitization for consistency.

IgorMinar reviewed (commented):
a few more comments / suggestions

mhevery commented on packages/core/src/sanitization/bypass.ts:
```diff
@@ -6,61 +6,120 @@
- * {@link TrustedStyleString}, {@link TrustedUrlString}
+ * @publicApi
  */
-export interface TrustedString extends String { [BRAND]: BypassType; }
```
I did not remove anything, I just moved it from `platform-browser` to the `core`. We could add additional branding if you would like, but that is not what we had in the past.

petebacondarwin replied:
Oh I see what you are doing... rather than use branded types you have switched to using real classes (e.g. `SafeValueImpl` and descendants). 

But in that case I don't see any value in having both real implementations and interfaces... especially since the interfaces are all semantically identically now...

```
const x: SafeValue = new SafeHtmlImpl('some string');
const y: SafeScript = x;  // No compile time error...
```

Why not just make them all classes and ditch the interfaces?

IgorMinar replied:
That would be a breaking change.

mhevery commented on packages/core/src/sanitization/bypass.ts:
```diff
@@ -6,61 +6,120 @@
+export function allowSanitizationBypassAndThrow(value: any, type: BypassType): boolean;
+export function allowSanitizationBypassAndThrow(value: any, type: BypassType): boolean {
+  const actualType = getSanitizationBypassType(value);
+  if (actualType != null && actualType !== type) {
```
No, I actually want this to be sensitive to `null` and `undefined`

petebacondarwin replied:
At least in the compiler we have been trying to be more explicit about these things, so if we wanted to compare both `null` and `undefined` then we would write it out explicitly using `===`. This would ensure people like @IgorMinar don't mistake this sort of comparison as an error. I think the minimizer might be able to collapse such comparisons back to `==` at bundling time so there is no benefit in using it in the source.

Alternatively, I think a comment explaining why `!=` is being used rather than` `!==` would help.

mhevery commented on packages/core/src/sanitization/bypass.ts:
```diff
@@ -6,61 +6,120 @@
+  const actualType = getSanitizationBypassType(value);
+  if (actualType != null && actualType !== type) {
+    // Allow ResourceURLs in URL contexts, they are strictly more trusted.
+    if (actualType == BypassType.ResourceUrl && type == BypassType.Url) return true;
```
fixed

mhevery commented on packages/platform-browser/src/security/dom_sanitization_service.ts:
```diff
@@ -154,78 +155,39 @@ export class DomSanitizerImpl extends DomSanitizer {
       case SecurityContext.SCRIPT:
-        if (value instanceof SafeScriptImpl) return value.changingThisBreaksApplicationSecurity;
-        this.checkNotSafeValue(value, 'Script');
+        if (allowSanitizationBypassAndThrow(value, BypassType.Script))
```
fixed

mhevery commented on packages/core/src/sanitization/sanitization.ts:
```diff
@@ -38,8 +38,8 @@ export function ɵɵsanitizeHtml(unsafeHtml: any): string {
   }
-  if (allowSanitizationBypass(unsafeHtml, BypassType.Html)) {
-    return unsafeHtml.toString();
+  if (allowSanitizationBypassAndThrow(unsafeHtml, BypassType.Html)) {
```
the function throws if the sanitization is not valid. Can you suggest a different name?

petebacondarwin replied:
Regarding the name, I feel that the problem is that this function is trying to do two different things:

1) check whether we can bypass the sanitizer for these types
2) check whether sanitization is allowed at all for these types

It seems to me that point 2) should be a check that the sanitizer should be doing e.g. in `_sanitizeHtml` and friends.

mhevery commented on packages/platform-browser/src/security/dom_sanitization_service.ts:
```diff
@@ -154,78 +155,39 @@ export class DomSanitizerImpl extends DomSanitizer {
-        if (value instanceof SafeScriptImpl) return value.changingThisBreaksApplicationSecurity;
-        this.checkNotSafeValue(value, 'Script');
+        if (allowSanitizationBypassAndThrow(value, BypassType.Script))
+          return unwrapSafeValue(value);
```
fixed

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

mhevery committed:
fixup! perf(core): make sanitization tree-shakable in Ivy mode

petebacondarwin commented on packages/core/src/sanitization/bypass.ts:
```diff
@@ -6,61 +6,120 @@
+    throw new Error(
+        `Required a safe ${type}, got a ${actualType} (see http://g.co/ng/security#xss)`);
+  }
+  return actualType === type;
```
So I think this function will:

* return `true` if
  * `actualType === type`
  * or `actualType` is `ResourceUrl` and `type` is `Url`
* return `false`: if
  * `actualType === null` and `type !== null`
  * or `actualType === undefined` and `type !== undefined`
* throw otherwise

Is that correct? It is not so obvious to glean that from reading the code.

mhevery replied:
Do you have suggestions as to make it more clear?

petebacondarwin commented on packages/core/test/render3/instructions_spec.ts:
```diff
@@ -161,33 +161,13 @@ describe('instructions', () => {
-        ɵɵstylingApply();
-      });
-      expect(sanitizerInterceptor.lastValue).toEqual(null);
-    });
```
Is this test no longer applicable? If it is then should it appear somewhere in the new code?

mhevery replied:
yes, this test is invalid, and has been removed.

petebacondarwin reviewed (commented):
I came to review for `docs-infra` but then found that you had removed all the changes in `aio` :-)

mhevery commented:
@petebacondarwin The type confusion comes from the fact that I am moving code between packages. There is no change to public API. Yes your suggestions would be a great improvement but it would be breaking change as it would make it more strict and so it should be done in separate PR.

mhevery commented:
[presubmit](http://test/OCL:263152058:BASE:263152086:1565713637340:c8799d)

IgorMinar reviewed (approved):
Lgtm, but please resolve Pete's comments.

mhevery commented:
[presubmit](http://test/OCL:263152058:BASE:263219519:1565732748631:fdd5ab75)

mhevery commented:
MERGE-ASSISTANCE: [Unrelated failure](http://test/OCL:263152058:BASE:263219519:1565732748631:fdd5ab75) (Unable to force green state)

AndrewKushnir commented:
The changes in this PR [break some g3 targets (Guitar tests) with Ivy](https://sponge.corp.google.com/invocation?id=df9be109-85ee-407c-b4bd-79de0970b695), adding "Blocked" label for now, so we can investigate problems before merging this PR.

AndrewKushnir commented:
Also started a [Global TAP Presubmit](https://test.corp.google.com/ui#id=OCL:263275270:BASE:263275368:1565755463975:e687484e) to see if there are other affected targets.

AndrewKushnir commented:
- [VE Presubmit](http://test/OCL:263458988:BASE:263459016:1565826584830:2b928306)
- [Ivy Presubmit](http://test/OCL:263458988:BASE:263459193:1565826644784:1e19f193)
